File: frontend/src/stream/events.ts
/**
 * The eight things that travel down a generation, as the browser types them —
 * and, from 2026-09-22, one more that only a live run sends.
 *
 * **The grammar is split in two, and the split is load-bearing.** The eight are
 * what a recording holds, line for line, and everything that reaches the map
 * comes from them. The ninth, `activity`, says what the model is doing inside a
 * call that has not come back yet; it is ephemeral by rule — never recorded,
 * never transcribed, never replayed, never invented by a replay — and it reaches
 * the run strip and nothing else (Kent, R44 and R47, 2026-09-22).
 *
 * Written by hand, exactly as `WorldView` was, because the server's description
 * of itself and this canvas are built at the same time. The server's own copy is
 * `backend/src/katalyst/engine/events.py`.
 *
 * **What pins the two together, and what cannot be pinned.**
 * `frontend/src/stream/__tests__/eventsMatchSchema.test-d.ts` fails the build
 * when this half and the generated `api/schema.ts` disagree — `tsc` is the
 * whole test, so there is nothing to run and nothing to remember to run.
 *
 * It cannot pin the eight **envelopes**, and nothing could: they travel as
 * server-sent events, and OpenAPI describes `text/event-stream` and stops. So
 * it pins them at the joints, which is where drift would actually hurt: every
 * piece an event is built from — `Proposition`, `Link`, `World`, `Violation`,
 * `Belief` — is aliased to the generated type below rather than retyped, so a
 * change to any of them is a compile error here without anybody writing a test.
 * And it pins the four shapes that do travel as JSON: the receipt, the request,
 * the drafted insert and the working.
 *
 * What is left is the eight names and which pieces each carries. The guard for
 * those is the running app — an event name this build does not know is counted
 * and shown (`growth.ts`), which turns a divergence into a number on screen
 * rather than into tiles that quietly never appear — and the server's own test
 * that the stream emits exactly those eight.
 *
 * **Field names are the wire's, in full.** `input_tokens`, not `inputTokens`. A
 * second spelling of a field is a second thing that can drift, and these come
 * straight off the stream and are printed straight onto the screen. The browser's
 * own view model, `WorldView`, keeps its own spellings because it is a different
 * thing with a different life.
 *
 * **Two facts the eight deliberately do not carry twice.** Whether a destination
 * was reached is on the verdict and nowhere else — the reason a *generation*
 * stopped is a different question with a different answer. And whether the run
 * was live or played back, the day a recording was made and the fingerprint of
 * the prompt are on the receipt and nowhere else. Two places to read one fact
 * eventually disagree.
 *
 * **One fact they do carry twice, and it is not the same mistake.** The set of
 * claims still open rides on both growth events. It is not one fact worked out
 * two ways; it is one fact restated after every change to it, so that whichever
 * event arrives is enough on its own to redraw the reserved rectangles.
 */

import type { components } from "../api/schema";
import type { Likelihood } from "../world/types";

/** An identifier the engine minted for a claim. Nothing in the browser ever makes one. */
export type PropositionId = string;

/** One claim, exactly as the rules layer writes one. */
export type Proposition = components["schemas"]["Proposition"];

/** One arrow, exactly as the rules layer writes one. */
export type Link = components["schemas"]["Link"];

/** One rule broken, with the rule's own code and the rule's own sentence. */
export type Violation = components["schemas"]["Violation"];

/** One finished world: a map with every likelihood worked through it. */
export type World = components["schemas"]["World"];

/** The run has started. The first thing that arrives, and it arrives at once. */
export interface GenerationStarted {
  readonly event: "generation_started";
  /** The engine's name for this run. What the transcript is asked for by. */
  readonly generation_id: string;
  /**
   * The seed every random draw in this run came from — always a number here,
   * even when the request left it out. Whoever chose it, this event is where the
   * browser reads it: the one the request sent, or the one the engine minted when
   * it sent none, or, on a replay, the one in the recording's header, which wins
   * over both. The browser prints what this says and never works out which case
   * it was.
   */
  readonly seed: number;
  /**
   * The same seed, exactly as it was written on the wire.
   *
   * **Put here by the reader, not by the server**, like the `event` field above
   * — and for a reason worth writing down. A seed is a whole number the engine
   * draws with, and it can be nineteen digits long; JavaScript holds a whole
   * number exactly only up to sixteen, so `seed` above is already rounded by the
   * time any code here sees it. The run this build was first watched on drew with
   * 4 803 646 386 380 448 080, and the browser's own reading of that ends 000.
   *
   * The browser never computes with a seed — it prints it, so that a reader can
   * ask for the same map again — so the honest thing to print is the digits that
   * came off the wire. Absent only when a reader handed the events over by hand
   * rather than reading them from a stream.
   */
  readonly seed_as_written?: string;
  /** The sentence the reader typed, in their own words. */
  readonly hypothesis: string;
  /** The Verify door's destination, in their words. Null when they used the Explore door. */
  readonly target: string | null;
}

/** One proposal the rules accepted. A claim, its incoming arrows, or both. */
export interface ProposalAccepted {
  readonly event: "proposal_accepted";
  /** Where this sits in the transcript, counting from 0. */
  readonly at: number;
  /** The new claim, with its identifier minted by the engine. Null for an arrows-only proposal. */
  readonly proposition: Proposition | null;
  /** Its incoming arrows, identifiers minted by the engine. */
  readonly links: readonly Link[];
  /** The claims still open to expand. What the reserved rectangles are drawn from. */
  readonly frontier: readonly PropositionId[];
}

/** One proposal the rules refused. An event, never an error. */
export interface ProposalRejected {
  readonly event: "proposal_rejected";
  readonly at: number;
  /** What the model wrote. Never minted, never drawn as a tile, only ever quoted. */
  readonly claim_in_words: string;
  /** The validator's own codes and its own sentences. One per rule broken. */
  readonly violations: readonly Violation[];
  /**
   * The claims still open to expand — the same field the accepted event carries,
   * and for the same reason. A claim closed by its third refusal in a row leaves
   * the frontier on this event, so its rectangle comes down at once instead of
   * standing there waiting for something that is not coming.
   */
  readonly frontier: readonly PropositionId[];
}

/** Every number, worked through the finished map, once. */
export interface BeliefsPropagated {
  readonly event: "beliefs_propagated";
  readonly world: World;
}

/** The Verify door's answer. Emitted only when the reader named a destination. */
export interface Verdict {
  readonly event: "verdict";
  readonly kind: "reached" | "no_path";
  /** The steps of the path, in order. Empty on `no_path`. */
  readonly path: readonly PropositionId[];
  /** The multiplied-out likelihood of those steps. Null on `no_path`. */
  readonly product: number | null;
  /** On `no_path`: the closest claim the map did reach. */
  readonly nearest: PropositionId | null;
  /** One plain sentence, the engine's own. */
  readonly why: string;
}

/** What the run cost. */
export interface Receipt {
  readonly event: "receipt";
  readonly model: string;
  readonly calls: number;
  readonly input_tokens: number;
  readonly output_tokens: number;
  readonly cache_read_tokens: number;
  /**
   * How many web searches this run made. Billed apart from tokens, so `dollars`
   * cannot be accounted for without it — which is why it is a field of its own
   * rather than something a reader is expected to infer.
   */
  readonly searches: number;
  readonly dollars: number;
  readonly seconds: number;
  /** Live, or played from a recording. The only place in the stream this is said. */
  readonly mode: "live" | "replay";
  /** The day the recording was made, as the map writes a day: `2026-09-18`. Null when live. */
  readonly recording_date: string | null;
  /**
   * How hard the model was asked to try, as a plain word.
   *
   * `default` when nothing was asked for and the service's own applied,
   * otherwise `low`, `medium`, `high`, `xhigh` or `max`. One setting with two
   * pinned defaults behind it: **a recording is made rich and a live run is
   * made fast** (Kent, G13, 2026-09-21), so two maps of the same sentence can
   * differ for a reason that has nothing to do with the sentence — and a reader
   * of a map is entitled to know which this was.
   *
   * It is a plain word rather than a number because it is the word the service
   * takes, and a number would be this browser translating one thing into
   * another and printing the translation.
   */
  readonly effort: string;
  /** A fingerprint of the prompt this run was made against. Always known. */
  readonly prompt_hash: string;
}

/** Why the generation stopped. Not whether the Verify door found anything — that is the verdict. */
export interface Done {
  readonly event: "done";
  readonly reason: StopReason;
  readonly claims: number;
  readonly links: number;
  readonly rejected: number;
}

/** The seven reasons a generation stops. There is no eighth, and none of them is `no_path`. */
export type StopReason =
  | "reached_terminal"
  | "depth_cap"
  | "width_cap"
  | "claim_cap"
  | "spend_cap"
  | "refusal_cap"
  | "no_terminal";

/** The run broke. One plain sentence, never a stack trace. */
export interface Failed {
  readonly event: "failed";
  readonly message: string;
}

/** Any one of the eight. The `event` field is what a `switch` walks. */
export type StreamEvent =
  | GenerationStarted
  | ProposalAccepted
  | ProposalRejected
  | BeliefsPropagated
  | Verdict
  | Receipt
  | Done
  | Failed;

/** The eight names, so a reader can tell one it knows from one it does not. */
export const EVENT_NAMES: readonly StreamEvent["event"][] = [
  "generation_started",
  "proposal_accepted",
  "proposal_rejected",
  "beliefs_propagated",
  "verdict",
  "receipt",
  "done",
  "failed",
];

/** Which of the three things the model was doing when it said so. */
export type ActivityKind = "searching" | "found" | "thinking";

/**
 * The ninth kind of line: what the model is doing **right now**.
 *
 * **It is not one of the eight, and the split is the whole point.** The eight
 * above are what a recording holds, line for line, and a replay is those eight
 * played back. This one is sent on a live run and is never written to a
 * recording, never written to a transcript, never replayed, never invented by a
 * replay, never counted by the `at` counter and never folded into the map. So a
 * recording is still the real stream line for line, which is what made a ninth
 * event possible at all: the thing it reports does not exist in a replay, the
 * same way the seconds counter's wait does not (Kent, R44 and R47, 2026-09-22).
 *
 * **What it carries is the model's own words and nothing composed by us.** The
 * search it just issued, verbatim; the title of one thing that search returned,
 * verbatim, with the host after it when a host is known; its own summarised
 * thinking, verbatim. There is no progress estimate here, no percentage and no
 * time remaining — the caps those would be worked out from are the server's and
 * are not on the wire.
 *
 * **Where it may be shown is the run strip and nowhere else.** Never on the map,
 * never in the panel, and never inside the polite live region: these lines
 * change every few seconds, and a region that read them out would be reading a
 * stopwatch over the top of the map being built.
 */
export interface Activity {
  readonly event: "activity";
  /**
   * The open claim the model call is working on — one of the identifiers in the
   * latest `frontier` — or null when the call is about no one claim, which is
   * the opening call that proposes the reader's own sentence as a claim.
   */
  readonly about: PropositionId | null;
  /** Which of the three things it is doing. */
  readonly kind: ActivityKind;
  /** The model's own words, verbatim. Never paraphrased, never composed here. */
  readonly text: string;
}

/** The ninth name. Live only, and a recording never holds it. */
export const ACTIVITY_NAME = "activity";

/**
 * The nine names a **live** run may send: the eight a recording holds, and the
 * ephemeral one.
 *
 * The reader matches on this list rather than on the eight, because a name that
 * is not on it comes back as *a name this build does not know* and is counted
 * on screen — which is the right answer for a server that has learned a tenth
 * word and the wrong one for a word this build knows perfectly well.
 */
export const NAMES_A_LIVE_RUN_MAY_SEND: readonly (StreamEvent | Activity)["event"][] = [
  ...EVENT_NAMES,
  ACTIVITY_NAME,
];

/**
 * An event whose name this build does not know.
 *
 * It is a shape of its own rather than a hole in the union: the reader has to
 * hand something back for every line on the wire, and a name with a payload
 * nobody can read is still a thing that happened. What is done with one is the
 * reducer's business — count it, change nothing else, and say so in the
 * transcript.
 */
export interface UnknownEvent {
  readonly event: "unknown";
  /** The name the wire carried. */
  readonly name: string;
  /**
   * True when the **name** was one of the eight and its payload could not be
   * read.
   *
   * Two different things end up here, and a screen that told a reader they were
   * the same would be telling them something false. *This build has no name for
   * that* is a browser older than its server, and the map it drew is a correct
   * map of the events it understood. *This build knows that name and could not
   * read what came with it* is a broken line on the wire, and the map may be
   * missing a claim — which is a different thing to be told, and a worse one.
   */
  readonly unreadable?: boolean;
}

/**
 * Everything the reader can hand back: one of the eight, the ephemeral ninth, or
 * a name it does not know.
 */
export type ReadEvent = StreamEvent | Activity | UnknownEvent;

/** What the browser asks for when it asks for a map. */
export interface GenerateRequest {
  /**
   * The sentence the reader typed — and, with no key, the whole of how the server
   * knows which recording to play. Required, and never empty.
   */
  readonly hypothesis: string;
  /** The Verify door's destination, in the reader's words. Absent is the Explore door. */
  readonly target?: string;
  /** The reader's own likelihood on the hypothesis. Absent when they chose "I don't know". */
  readonly user_belief?: Likelihood;
  /**
   * Only when reproducing a run we were handed a seed for.
   *
   * **The browser does not invent one.** With none, the engine mints one and the
   * first event says which, so the run is reproducible from the moment it starts
   * and the number on screen is still the engine's.
   */
  readonly seed?: number;
  /**
   * How this run starts: `live` calls a model, `replay` plays the committed
   * recording of this sentence back.
   *
   * **The caller says which, and the server never chooses** (record 0012,
   * amended 2026-09-21). Left out, the server plays a recording, because a
   * request that did not ask to spend money must never spend it.
   */
  readonly start?: "replay" | "live";
}

File: backend/src/katalyst/engine/events.py
"""The eight things a generation is made of, one that is only ever live, and their framing.

One request builds a whole map, and this is what comes back along it while that
happens. **The stream is the loading state**: a generation takes minutes and
returns one whole proposal every few seconds, so there is no honest way to show
it as a single answer that arrives at the end — and a spinner followed by a
finished map is a veto condition rather than a design choice.

Eight names that are written down
---------------------------------
The map starts · a proposal landed · a proposal was refused · the likelihoods ·
whether the destination was reached · what it cost · why it stopped · what broke.

**These eight are the generation.** Each one is a decision somebody took — the
model, the rules, the walk — and a recording is exactly these, line for line.

Two things that close a claim make **no event of their own**: the model saying
this part of the story is finished, and a claim's third failure in a row. Neither
adds anything to the map, and the browser does not need to be told — it reads the
next growth event's `frontier`, which no longer names that claim, and on
`beliefs_propagated` every skeleton goes at once because the frontier is empty by
definition. An event whose only job is to say that nothing happened is an event
somebody will one day draw.

And one that never is
---------------------
`activity` says what a model call is doing **right now** — the search it just
issued, one thing that search returned, the sentence it is thinking. It is not a
decision and it changes nothing, so it is not part of the generation: it is sent
on a live run and **never written to a recording, a transcript or a kept run,
never replayed, never invented by a replay, never billed and never counted by
`at`**. That is why it is not in `NAMES` and not in `Event` — the eight are what
a recording holds, and the ninth is what a live run additionally says out loud
(record 0027, Kent's R44 and R47, 2026-09-22; it amends record 0023's *no ninth
event*).

Four of these shapes are not defined here, on purpose
------------------------------------------------------
`Verdict` is the Verify door's own answer and `Done.reason` is the walk's own
list of reasons; both are imported rather than restated, because a shape written
out twice is two shapes that eventually disagree. `Proposition`, `Link`,
`Violation` and `World` are the rules layer's, carried whole, so what the browser
draws is what `domain/` computed.

What this file must never do
----------------------------
- Never gain another event for something a client can already read off the eight.
- Never carry a fact that is already on another event. `no_path` lives on the
  verdict; the mode, the recording's date and the prompt's fingerprint live on
  the receipt. Two places to read one fact eventually disagree.
- Never write a stack trace, an exception's name or an identifier into `Failed`.
  It is interface text.
- Never put `activity` in `NAMES`, in `BY_NAME` or in `Event`. Those three are
  what a recording is read and written through, and the moment `activity` is in
  one of them a recording can hold one.
- Never emit a heartbeat, a comment line or anything else that is not one of the
  eight or `activity`. A recording is the eight line for line, and a line
  carrying no event is a line somebody eventually parses as one.
"""

import json
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from katalyst.domain import Link, Proposition, PropositionId, Violation, World
from katalyst.engine.grow import StoppingReason
from katalyst.engine.outcome import (
    WHAT_THE_SERVICE_DEFAULTS_TO,
    Accepted,
    Outcome,
    Refused,
)
from katalyst.engine.verify import Verdict


class GenerationStarted(BaseModel):
    """The generation has an identifier and a seed. Nothing has been proposed yet.

    It goes out before the first model call has returned, which is what the first
    screen draws its reserved rectangle from: the first thing a person sees is the
    map beginning, not a word of the model's.
    """

    model_config = ConfigDict(frozen=True)

    generation_id: str = Field(description="Minted by `engine/ids.py`, never by the model.")
    seed: int = Field(
        description=(
            "The one number the rules layer will work every likelihood through with. "
            "Always a number here, even when the request left it out — the server "
            "mints one and says which, so the run is reproducible from its first event."
        )
    )
    hypothesis: str = Field(description="The sentence the person typed, unaltered.")
    target: str | None = Field(
        default=None, description="The Verify door's destination, in their own words."
    )


class ProposalAccepted(BaseModel):
    """One proposal passed the rules. The map is now bigger by this much.

    Two different proposals arrive through it. One that adds a claim carries the
    claim and its one incoming arrow; one that only joins two claims already on
    the map carries no claim at all and one new arrow. From the browser's side
    both are the same fact — the map got bigger — and splitting them would be a
    ninth event whose only difference is a field that is already optional.

    The claim a person's own sentence became arrives this way too, with no arrows,
    because it has no cause. Turning a sentence into something checkable is
    exactly the step a reader wants to audit.
    """

    model_config = ConfigDict(frozen=True)

    at: int = Field(
        description=(
            "Where this sits in the transcript, counting from 0. It rises by one on "
            "every accepted and every refused proposal, because both are things that "
            "were proposed. It is not a time."
        )
    )
    proposition: Proposition | None = Field(
        default=None,
        description="The new claim, with the identifier minted here. Absent for an arrow alone.",
    )
    links: tuple[Link, ...] = Field(
        default=(), description="Its incoming arrows, with the identifiers minted here."
    )
    frontier: tuple[PropositionId, ...] = Field(
        default=(), description="The claims still open to expand, once this had been folded in."
    )


class ProposalRejected(BaseModel):
    """One proposal was refused. Every reason at once, never the first one only.

    It is an event, not an error: the answer stays 200 and the stream carries on.
    Watching the rules refuse the model is half of what this product is for, and a
    generation that hides its misses has deleted that half.
    """

    model_config = ConfigDict(frozen=True)

    at: int = Field(description="The same transcript counter the accepted proposals use.")
    claim_in_words: str = Field(
        description=(
            "What the model wrote, or the plain sentence of what happened when it "
            "wrote nothing we could read. Never minted, never drawn as a tile."
        )
    )
    violations: tuple[Violation, ...] = Field(
        default=(),
        description=(
            "Every reason the map's own rules gave, in their words. Empty when no map "
            "was proposed at all — when the model declined, or its answer did not fit."
        ),
    )
    frontier: tuple[PropositionId, ...] = Field(
        default=(),
        description=(
            "The claims still open. **Both** growth events carry it, so a claim closed "
            "by its third failure loses its reserved rectangle at once rather than "
            "leaving one standing where nothing will arrive."
        ),
    )


class BeliefsPropagated(BaseModel):
    """Every likelihood, worked through the finished map. Exactly once, at the end.

    Never during the growing. A chip that changes four times as its causes arrive
    has shown four numbers nobody computed, and only the last of them is the
    engine's answer.
    """

    model_config = ConfigDict(frozen=True)

    world: World = Field(description="The whole thing, straight from the rules layer.")


class Receipt(BaseModel):
    """What this run cost. The second-to-last event of every stream.

    Second-to-last in both endings — before the run finished and before it broke.
    A run that broke after ten calls still cost ten calls, and a run that broke
    before its first emits a receipt of zeroes, which is also true.

    It is not the same shape as the running total `engine/receipt.py` keeps. That
    one counts what the bill is made of, cache writes included; this one is what a
    person is shown, and it carries the three facts about the run that live here
    and nowhere else: whether it was live or played back, the day the recording
    was made, and the fingerprint of the prompt behind it.
    """

    model_config = ConfigDict(frozen=True)

    model: str = Field(description="Which model wrote this map.")
    calls: int = Field(description="How many round trips the run made.")
    input_tokens: int = Field(description="Tokens of question read fresh.")
    output_tokens: int = Field(description="Tokens of answer written, thinking included.")
    cache_read_tokens: int = Field(
        description=(
            "Tokens the service recognised from an earlier call. A run where this "
            "stays at nothing is a bug in how the request is put together."
        )
    )
    searches: int = Field(
        description=(
            "Web searches this run made. Billed apart from tokens, so the dollars "
            "below cannot be worked out again without it."
        )
    )
    dollars: float = Field(description="What all of that came to, at the shipped price table.")
    seconds: float = Field(description="How long the run took, wall clock.")
    mode: Literal["live", "replay"] = Field(
        description="Whether this run called a model or played a recording back."
    )
    recording_date: date | None = Field(
        default=None, description="The day the recording was made. Nothing at all when live."
    )
    effort: str = Field(
        default=WHAT_THE_SERVICE_DEFAULTS_TO,
        description=(
            "How hard the model was asked to try, as a plain word — `default` "
            "when nothing was sent and the service's own applied, otherwise "
            "`low`, `medium`, `high`, `xhigh` or `max`. One setting with two "
            "pinned defaults behind it: a recording is made rich and a live run "
            "is made fast, and a reader of a map is entitled to know which this "
            "was (Kent, G13, 2026-09-21)."
        ),
    )
    prompt_hash: str = Field(description="The fingerprint of the prompt this run was made against.")


class Done(BaseModel):
    """The generation finished. One reason, and the size of what was built."""

    model_config = ConfigDict(frozen=True)

    reason: StoppingReason = Field(
        description=(
            "Why the generation stopped. One reason, picked by the walk's own rule: it "
            "names what closed the last claim that was still open, with the money "
            "running out and a map that ends nowhere as the two overrides above it."
        )
    )
    claims: int = Field(description="How many claims the map ended up with.")
    links: int = Field(description="How many arrows.")
    rejected: int = Field(description="How many proposals were refused along the way.")


class Failed(BaseModel):
    """Something on our side broke. One plain sentence, never a stack trace.

    Stopping because the money ran out is **not** this: that is a decision, and it
    arrives as `done` with its own reason. This is for the things nobody chose.
    """

    model_config = ConfigDict(frozen=True)

    message: str = Field(
        description=(
            "One plain sentence a person can act on. Never an exception's name, never "
            "an identifier, never a stack trace — those go in the server's own log."
        )
    )


class Activity(BaseModel):
    """What one model call is doing this second. Live only, and never written down.

    The one thing on the wire that is **not** a decision. Nothing on the map
    changes because of it, nothing is spent because of it, and a run that never
    sent one is the same run. It exists because a model call takes minutes and a
    person watching an empty screen has no way to tell a working program from a
    stopped one (Kent, 2026-09-22).

    Every word in `text` is the model's own or the search tool's own, verbatim.
    Nothing here is composed by us, and nothing here is an estimate of progress:
    there is no percentage, no count of what is left and no guess at how long.

    What it must never do
    ---------------------
    - Never be written to a recording, a transcript or a kept run.
    - Never be sent by a replay, and never be invented by one.
    - Never be billed, and never counted by the `at` counter.
    - Never be folded into the map, and never drawn on it.
    - Never be the reason a run fails. Whatever goes wrong raising one is
      swallowed and logged, and the run carries on.
    """

    model_config = ConfigDict(frozen=True)

    about: PropositionId | None = Field(
        default=None,
        description=(
            "The open claim this call is working on — one of the identifiers the "
            "latest `frontier` names — or nothing at all when the call is not "
            "about one claim, which is what the opening call is."
        ),
    )
    kind: Literal["searching", "found", "thinking"] = Field(
        description=(
            "Which of the three this is. `searching`: the model just issued a web "
            "search. `found`: one thing that search returned. `thinking`: the "
            "model's own summarised thinking."
        )
    )
    text: str = Field(
        description=(
            "The words themselves, never paraphrased by us. For `searching`, the "
            "query the model wrote. For `found`, the title of one page the search "
            "returned, followed by ` · <host>` when a host is known. For "
            "`thinking`, the most recent whole sentence of the model's own "
            "summarised thinking."
        )
    )


ACTIVITY = "activity"
"""The name the ninth travels under.

**Deliberately not in `NAMES`.** That table is the eight a recording is written
and read through, and a name in it is a name a recording may hold.
"""


Event = (
    GenerationStarted
    | ProposalAccepted
    | ProposalRejected
    | BeliefsPropagated
    | Verdict
    | Receipt
    | Done
    | Failed
)
"""Any one of the eight. The wire's `event:` line carries the name; `data:` carries the payload.

`Activity` is **not** one of them, on purpose: this is the type a recording, a
transcript and a replay are written in terms of.
"""

OnTheWire = Event | Activity
"""Anything a live stream may write out: the eight, plus the one that is never recorded."""

NAMES: dict[type[Event], str] = {
    GenerationStarted: "generation_started",
    ProposalAccepted: "proposal_accepted",
    ProposalRejected: "proposal_rejected",
    BeliefsPropagated: "beliefs_propagated",
    Verdict: "verdict",
    Receipt: "receipt",
    Done: "done",
    Failed: "failed",
}
"""Every event and the name it travels under. **The one place the eight are listed.**

Written as a table rather than as a field on each shape, because one of the eight
— the Verify door's answer — is defined where that door lives, and a shape should
not grow a field in order to be streamable.
"""

BY_NAME: dict[str, type[Event]] = {said: shape for shape, said in NAMES.items()}
"""The same table read the other way, for whatever reads a stream back."""


def framed(event: OnTheWire) -> str:
    """Write one event the way it goes on the wire: two lines and a blank one.

    The payload is **one line of JSON** — never pretty-printed, never wrapped — so
    a reader never has to join two `data:` lines back together and a recording can
    be one event per line.

    Args:
        event: The event to write. One of the eight, or the live-only ninth.

    Returns:
        The `event:` line, the `data:` line and the blank line that ends them.
    """
    return f"event: {name_of(event)}\ndata: {payload_of(event)}\n\n"


def name_of(event: OnTheWire) -> str:
    """Say which name this event travels under: one of the eight, or `activity`."""
    if isinstance(event, Activity):
        return ACTIVITY
    return NAMES[type(event)]


def payload_of(event: OnTheWire) -> str:
    """Write one event's payload as the single line of JSON the wire carries.

    Sorted and tightly spaced, so the same event is always the same bytes and a
    recording can be compared with a live run without either being re-read.
    """
    return json.dumps(json.loads(event.model_dump_json()), separators=(",", ":"), sort_keys=True)


def growth_event(outcome: Outcome, at: int) -> ProposalAccepted | ProposalRejected:
    """Turn one folded outcome into the event that says what happened to it.

    The one place an outcome becomes something on the wire, so the stream and the
    recorder cannot drift into saying it two ways.

    Args:
        outcome: What happened to one proposal, and what it cost.
        at: Where it sits in the transcript, counting from 0.

    Returns:
        The growth event. A model that had nothing more to say makes none of
        these, and the caller checks that before asking.
    """
    result = outcome.result
    if isinstance(result, Accepted):
        return ProposalAccepted(
            at=at, proposition=result.proposition, links=result.links, frontier=outcome.frontier
        )
    if not isinstance(result, Refused):  # pragma: no cover - `makes_an_event` decided this
        raise TypeError(
            "growth_event was handed an answer that makes no event. `makes_an_event` "
            "says which do, and the caller asks it first."
        )
    return ProposalRejected(
        at=at,
        claim_in_words=result.claim_in_words,
        violations=result.violations,
        frontier=outcome.frontier,
    )
